pkg/agent/clientset.go

```diff
@@ -211,20 +217,32 @@ func (cs *ClientSet) sync() {
 	}
 }
 
-func (cs *ClientSet) connectOnce() error {
-	if !cs.syncForever && cs.serverCount != 0 && cs.ClientsCount() >= cs.serverCount {
+func (cs *ClientSet) ServerCount(ctx context.Context) int {
+	var serverCount int
+	if cs.serverCounter != nil {
+		serverCount = cs.serverCounter.Count(ctx)
+		if serverCount == 0 {
+			klog.Warningf("server lease counter could not find any leases")
+		}
+	} else {
+		serverCount = cs.lastReceivedServerCount
+	}
+
+	metrics.Metrics.SetServerCount(serverCount)
+	return serverCount
+}
+
+func (cs *ClientSet) connectOnce(ctx context.Context) error {
+	serverCount := cs.ServerCount(ctx)
+
+	if !cs.syncForever && serverCount != 0 && cs.ClientsCount() >= serverCount {
```

Comment:
ServerCount can currently return 0. If that happens consistently across agents we can wedge the cluster.

Reply:
I thought that's what the `serverCount != 0` here is trying to prevent? It's saying that even if `ClientsCount() == serverCount` then if the `serverCount == 0` you should keep trying to connect anyway.

Reply:
Lease count always returns at least 1 now.

Reply:
Should we remove the `serverCount != 0` check then?